cpp/src/parquet/column_reader.cc

```diff
@@ -1370,6 +1402,54 @@ class TypedRecordReader : public TypedColumnReaderImpl<DType>,
     return bytes_for_values;
   }
 
+  // Two parts different from original HasNextInternal:
```

Comment:
What would a `SkipRecords(k)` do? During `SkipRecords`, `HasNext` might also being called, would a hidden `HasNextInternal()` skip unexpected rows?

Reply:
- Within a page, the hit rows may not be contiguous. For a page that has already been hit, any non-hit rows within it need to be skipped using ``SkipRecords``.  It may be clearer to look at this logic in conjunction with code comments ``column_reader.h:164~173``:
```
//
  //   | <--------------------------- column chunk -------------------------------> |
  //             | <-------------------- page N -----------------------> |
  //        first_row_idx                                          last_row_idx
  //   |-- ... --|-------------------------------------------------------|--- ... ---|
  //                       |---- range0 ----|         |---- range1 ----|
  //             |--skip0--|                |--skip1--|
  //             |------last_row_index0-----|
  //             |-------------------last_row_index1-------------------|
  //
```
- During the execution of the ``SkipRecords`` method, it does indeed call the ``HasNextInternal`` method. The current implementation and testing have covered such scenarios, ensuring that no errors occur.

Reply:
Yeah, I mean it limits the usage of `TypedColumnReader`, and only allow internal skip. External skip would introduce inconsistent skipping

Reply:
Oh I get to understand this. `SkipRecords()` now means "trying to skip the existing records and skip any remaing data"...

Reply:
First of all, it is strange to execute ``SkipRecords`` on the basis of hit lines. Secondly, whether it is consistent depends on how to understand the semantics of ``SkipRecords``. If some lines are skipped on the basis of hit lines, the current implementation can theoretically guarantee consistency, but more tests need to be added for verification; If ``SkipRecords`` is for all rows in page, then the existing implementation will indeed have problems.

Reply:
> Yeah, I mean it limits the usage of `TypedColumnReader`, and only allow internal skip. External skip would introduce inconsistent skipping. In current-case, `Skip(skip_last)` would skip more than `skip_last`.

First of all, it is strange to execute ``SkipRecords`` on the basis of hit lines. Secondly, whether it is consistent depends on how to understand the semantics of ``SkipRecords``. If some lines are skipped on the basis of hit lines, the current implementation can theoretically guarantee consistency, but more tests need to be added for verification; If ``SkipRecords`` is for all rows in page, then the existing implementation will indeed have problems.

Reply:
> First of all, it is strange to execute `SkipRecords` on the basis of hit lines. 

I think if Parquet user is using a dense read with selection vector, or doing something like incremental filtering, it's possible to do this. Since `SkipRecords` is exported, and it's used when skipping the records within page, I think the interface is a bit weird🤔

Reply:
> > First of all, it is strange to execute `SkipRecords` on the basis of hit lines.
> 
> I think if Parquet user is using a dense read with selection vector, or doing something like incremental filtering, it's possible to do this. Since `SkipRecords` is exported, and it's used when skipping the records within page, I think the interface is a bit weird🤔

I get your point, maybe this is just a code implementation problem, we can perform row skipping operations in ``ReadRecords``. What do you think?